docs/fp-core.md

```diff
@@ -0,0 +1,90 @@
+
+## Invariants
+
+We define invariants that the finality provider should satisfy:
+
+- Finality signature should be submitted for a block if:
+  - the block is not finalized,
+  - the finality provider has positive voting power for this block,
+  - the finality provider has not cast a finality signature for this block, and
+  - the finality provider has cast finality votes for all non-finalized blocks
+    prior to this block for which it has voting power.
+
+- Public randomness should be committed for a block if:
+  - the block is not finalized, and
+  - the finality provider has not committed any randomness for this block
+
+- A finality provider instance should not be running if it’s status is `slashed`
+  or `jailed` on Babylon.
+
+#### Internal variables
+
+The finality provider maintains the following internal variables.
+* `last_voted_height`: the last block height on which the fp has cast a vote,
+* `last_processed_height`: the last block height on which the fp either has
+  cast a vote or it does not have voting power, and
+* `next_height`: the next block to poll.
+
+Internal variables need to satisfy the following rules:
+* `last_processed_height >= last_voted_height` should always hold, and
+* `next_height > max{latest_finalized_height, last_processed_height}` should
+  always hold.
+
+## Heuristics
+
+### Sending finality votes
+
+* The poller polls blocks from start_height one by one monotonically.
+* `start_height` should be the least height that satisfy
+  `start_height > max{latest_finalized_height, last_processed_height}`.
+* For the next block from the poller, the finality provider retries to send
+  a finality signature until the invariant is not satisfied.
+
+### Committing public randomness
+
+* Finality providers should start to send public randomness commit right
+  after it is registered on Babylon. This is because a public randomness commit
+  only takes into effect after the committed epoch is finalized. So the finality
+  provider should start sending the commit as soon as possible.
+* If the remaining randomness is not sufficient (defined by config) by the
+  current babylon height, the finality provider should make a new commit.
+* The next commit should be made with `start_height` right after the last
+  committed height. This is to ensure no gap between each commit.
+
+The above approach of committing public randomness may cause waste of some
+public randomness in cases where the finality provider does not have voting
+power for some heights, but given that the randomness are constructed in a
+merkle-tree and the commit only contains the root hash of the tree. Therefore,
+the waste is negligible.
+
+### Fast sync
+
+Fast sync is needed for bootstrap and in case the finality provider falls
+behind due to network latency.
+
+* The finality provider enters fast sync if:
+  * the finality provider is bootstrapping, or
+  * the `current_block_height - the last processed height > gap` (configured value).
```

Comment:
After https://github.com/babylonlabs-io/finality-provider/pull/57, FP in the normal-case also batch submits finalkity signatures. Does it mean that the normal-case execution is same as fast sync with smaller gaps? If yes, does it make sense to harmoise the normal-case execution and fastsync, or even remove the fast sync completely?

Reply:
Possibly we can remove the case where the fp enters fast-sync from normal-execution but we still need fast sync for bootstrap. I'll keep it for now

Reply:
I see, sg